src/caffe/layers/spp_layer.cpp

```diff
@@ -0,0 +1,195 @@
+
+  bottom_h_ = bottom[0]->height();
+  bottom_w_ = bottom[0]->width();
+  CHECK_GT(bottom_h_, 0) << "Input dimensions cannot be zero.";
+  CHECK_GT(bottom_w_, 0) << "Input dimensions cannot be zero.";
+
+  pyramid_height_ = spp_param.pyramid_height();
+  split_top_vec_.clear();
+  pooling_bottom_vecs_.clear();
+  pooling_layers_.clear();
+  pooling_top_vecs_.clear();
+  pooling_outputs_.clear();
+  flatten_layers_.clear();
+  flatten_top_vecs_.clear();
+  flatten_outputs_.clear();
+  concat_bottom_vec_.clear();
+
+  // split layer output holders setup
+  for (int i = 0; i < pyramid_height_; i++) {
+    split_top_vec_.push_back(new Blob<Dtype>());
+  }
+
+  // split layer setup
+  LayerParameter split_param;
+  split_layer_.reset(new SplitLayer<Dtype>(split_param));
+  split_layer_->SetUp(bottom, split_top_vec_);
+
+  for (int i = 0; i < pyramid_height_; i++) {
+    // pooling layer input holders setup
+    pooling_bottom_vecs_.push_back(new vector<Blob<Dtype>*>);
+    pooling_bottom_vecs_[i]->push_back(split_top_vec_[i]);
+
+    // pooling layer output holders setup
+    pooling_outputs_.push_back(new Blob<Dtype>());
+    pooling_top_vecs_.push_back(new vector<Blob<Dtype>*>);
+    pooling_top_vecs_[i]->push_back(pooling_outputs_[i]);
+
+    // pooling layer setup
+    LayerParameter pooling_param = GetPoolingParam(
+        i, bottom_h_, bottom_w_, spp_param);
+
+    pooling_layers_.push_back(new PoolingLayer<Dtype>(pooling_param));
+    pooling_layers_[i]->SetUp(*pooling_bottom_vecs_[i], *pooling_top_vecs_[i]);
+
+    // flatten layer output holders setup
+    flatten_outputs_.push_back(new Blob<Dtype>());
+    flatten_top_vecs_.push_back(new vector<Blob<Dtype>*>);
+    flatten_top_vecs_[i]->push_back(flatten_outputs_[i]);
+
+    // flatten layer setup
+    LayerParameter flatten_param;
+    flatten_layers_.push_back(new FlattenLayer<Dtype>(flatten_param));
+    flatten_layers_[i]->SetUp(*pooling_top_vecs_[i], *flatten_top_vecs_[i]);
+
+    // concat layer input holders setup
+    concat_bottom_vec_.push_back(flatten_outputs_[i]);
+  }
+
+  // concat layer setup
+  LayerParameter concat_param;
+  concat_layer_.reset(new ConcatLayer<Dtype>(concat_param));
+  concat_layer_->SetUp(concat_bottom_vec_, top);
+}
+
+template <typename Dtype>
+void SPPLayer<Dtype>::Reshape(const vector<Blob<Dtype>*>& bottom,
+      const vector<Blob<Dtype>*>& top) {
+  CHECK_EQ(4, bottom[0]->num_axes()) << "Input must have 4 axes, "
+      << "corresponding to (num, channels, height, width)";
+  channels_ = bottom[0]->channels();
+  bottom_h_ = bottom[0]->height();
+  bottom_w_ = bottom[0]->width();
+  SPPParameter spp_param = this->layer_param_.spp_param();
+  split_layer_->Reshape(bottom, split_top_vec_);
+  for (int i = 0; i < pyramid_height_; i++) {
+    LayerParameter pooling_param = GetPoolingParam(
+        i, bottom_h_, bottom_w_, spp_param);
+
+    delete pooling_layers_[i];
+    pooling_layers_[i] = new PoolingLayer<Dtype>(pooling_param);
```

Comment:
Can you make `pooling_layers_` a `vector<shared_ptr<PoolingLayer<Dtype> > >` and change the above two lines to just `pooling_layers_[i].reset(new PoolingLayer...)`? I don't think the last set of `PoolingLayer`s you create here will be deleted with your current implementation.